apps/server/src/services/rollUtils.ts

```diff
@@ -153,6 +154,7 @@ type CurrentTimers = {
   selectedEventId: string | null;
   current: number | null;
   _finishAt: number | null;
+  _startAt: number | null;
```

Comment:
I might be wrong, but I dont think we need to get this far to catch a skip.
in the `update` method of the `TimerService` we have the clock

before setting the new clock, we can check if the timer has skipped. 
we already have the current logic 
```ts
    const previousTime = this.timer.clock;
    this.timer.clock = clock.timeNow();
    if (previousTime > this.timer.clock) {
      force = true;
    }
```

which we can modify to include our logic (this is just an example and will likely not work, we would want something more intelligent)
```ts
    const hasSkipped = Math.abs(previousTime - this.timer.clock) > config.rollSkipLimit;
    if (hasSkipped) {
      // maybe we can even check here if the skip is larger than the duration of the event?
      // we can skip within the event
      PlaybackService.roll();
      force = true;
    }
```

Reply:
we could 
but it looks a bit like we would be duplicating  almost all the functionality of  `updateRoll` to know if we skipped out of an event 
is that preferable? or maybe empty out `updateRoll`

Reply:
I am unsure on why this would be the case.
The code above likely works. the optimisation I mentioned about the skipping would also only be a few lines of code

like
```ts
    const skipTime = previousTime - this.timer.clock;
    const hasSkipped = Math.abs(skipTime) > config.rollSkipLimit;

    if (hasSkipped) {
      if (skipTime > 0 && skipTime > this.timer.current) {
        // we skipped out of the event to the next one
        PlaybackService.roll();
      } else if (skipTime < 0 && skipTime < this.timer.elapsed) {
        // we might have skipped outside of the event
        PlaybackService.roll();
      }

      // otherwise we just skipped within the event
      force = true;
    }
```

I may be wrong, in which case feel free to ignore. We definitely dont want to duplicate the logic of updateRoll here

Reply:
can we write tests for `TimerService.ts`? That is the good thin about having it in `rollUtils.ts`
it would end up somthing like this right?
```typescript
//TimerService.ts
  update(force = false) {
    const previousTime = this.timer.clock;
    this.timer.clock = clock.timeNow();
    if (previousTime > this.timer.clock) {
      force = true;
    }
    const skipTime = previousTime - this.timer.clock;
    const hasSkipped = Math.abs(skipTime) > config.rollSkipLimit;

    // we call integrations if we update timers
    let shouldNotify = false;
    if (this.playback === Playback.Roll) {
      shouldNotify = true;
      if (hasSkipped) {
        if (skipTime > 0 && skipTime > this.timer.current) {
          // we skipped out of the event to the next one
          PlaybackService.roll();
        } else if (skipTime < 0 && skipTime < this.timer.elapsed) { 
          // we might have skipped outside of the event
          PlaybackService.roll();
        } else {
          // otherwise we just skipped within the event
          this.updateRoll();
        }
      } else {
        this.updateRoll();
      }
    } else if (this.timer.startedAt !== null) {
      // we only update timer if a timer has been started
      shouldNotify = true;
      this.updatePlay();
    }
``` 
am I wrong in thinking that `} else if (skipTime < 0 && skipTime < this.timer.elapsed) { ` 
is meant to check if we have skipped backwards out of an event
and should be `} else if (skipTime < 0 && Math.abs(skipTime) > this.timer.elapsed) { `

Reply:
we can absolutely write tests and you are encouraged to do so.
having said that, the current architecture of this service makes it hard to test. I am hoping to change some of that during the new work.

For now we extract some of the core calculations in `rollUtils` and `timerUtils` for this exact issue.

As for the logic, yes, you are correct.
skip time can be negative, but the elapsed time can only ever be positive
Keep in mind the code I wrote here was just a quick and rough example, it is untested and I did not considered any edge cases